### ISSUE TITLE
Task List

### ISSUE THREAD

**Issue by nilsleh** [lightning-uq-box/lightning-uq-box#23]
This is a list of tasks encountered while writing code about changes / simplifications that should be made at some point. If anyone has time or wants to take on any of the tasks, you are welcome to do so. Then you can open a PR and link it in this issue so we can keep track.

- [x] several implemented uq_methods during the predict step have an if/else statement depending on nll/mse prediction. This is a lot of code duplication and a single utility function should be written that can be used by models to remove code duplication (#25 )
- [ ] We should implement NLL loss with pytorch implemented [GaussianNLLLoss](https://pytorch.org/docs/stable/generated/torch.nn.GaussianNLLLoss.html) or alternatively, everything should be with our custom NLL log_sigma function with clear motivation why we need it (in documentation)
- [x] [NEP29](https://numpy.org/neps/nep-0029-deprecation_policy.html#support-table) has dropped support for python 3.8. Since many libraries like Pytorch follow this scheme, we should also only support python 3.9. This will introduce some style and typing changes that can easily be handled with [pyupgrade](https://github.com/asottile/pyupgrade). python 3.9 has native type hint support which is nice (#26)
- [x] change the configuration files and initialization by using [hydra](https://github.com/facebookresearch/hydra), for an example see [this](https://hydra.cc/docs/advanced/instantiate_objects/overview/). since it allows even nested instantiating of classes, it seems like a perfect fit for our UQ methods where we want to ensure reproducibility by defining everything from configs. This will make the config structure much clearer and we don't have to come up with our own nested config keys (#27)

Documentation:
- [ ] the `testing_uq_method.py` files we have created to test our implementations should become their own tutorial notebooks. The idea would be to convert the python file to a notebook so that we can also write explanations/derivations/practical tips etc about a method and at the same time illustrate how to use this library
- [ ] tutorials should include a note that explains the available hyperparamters unique to this method
- [ ] Categorize the methods for the documentation page
- [ ] Check for any documentation/argument inconsistencies in the common methods like `training_step`, `test_step`, or `predict_step`
- [ ] for every method, if we are implementing certain equations from papers, we should have comments saying that. The goal of the library is also to make it easier for others to understand UQ-methods and clear comments are definitely helpful to link theory and code

Testing:
- [ ] expand test coverage: basic pytest files are setup for each method, but not with 100% test coverage, which we need to aim for (#36)
- [ ] implement [mypy](https://github.com/python/mypy) static type hinting checks: this will require some work but is a great tool for bug checking and something that well maintained software projects have (#37)
